src/main/scala/com/fulcrumgenomics/commons/util/Async.scala

```diff
@@ -0,0 +1,286 @@
+      * or [[RuntimeException]] as appropriate.
+      */
+    private final def checkAndRethrow(): Unit = {
+      this.throwable.getAndSet(None).foreach { thr =>
+        this.isClosed.set(true) // Ensure no further attempts to write
+        thr match {
+          case e: Error             => throw e
+          case ex: RuntimeException => throw ex
+          case t                    => throw new RuntimeException(t)
+        }
+      }
+    }
+
+    /** A [[Runnable]] implementation that reads from the blocking queue and processes the item with the [[sink]]
+      * method.
+      */
+    private class AsyncSinkRunnable extends Runnable {
+      override def run(): Unit = {
+        try {
+          // The order of the two conditions is important, because we want to make sure that emptiness status of the
+          // queue does not change after we have evaluated isClosed as it is now (isClosed checked before
+          // queue.isEmpty), the two operations are effectively atomic if isClosed returns true
+          while (!outer.isClosed.get || !outer.queue.isEmpty) {
+            val item = outer.queue.poll(50, TimeUnit.MILLISECONDS)
+            if (item != null) outer.sink(item)
+          }
+        } catch  {
+          case thr: Throwable =>
+            outer.throwable.compareAndSet(None, Some(thr))
+
+            // In case the sink was blocking on a full queue before ex has been set, clear the queue
+            // so that the sink will no longer be blocked so that it can see the exception.
+            outer.queue.clear()
+        }
+      }
+    }
+  }
+
+  /** An asynchronous wrapper around a [[Writer]] class.
+    *
+    * @param writer the writer to wrap
+    * @param bufferSize the number of elements to buffer before blocking when writing the elements
+    * @tparam T the type of object to write
+    */
+  class AsyncWriter[T](val writer: Writer[T], bufferSize: Int = DefaultBufferSize)
+    extends AsyncSink[T](sink=writer.write, bufferSize=bufferSize, source=Some(writer)) with Writer[T] {
+    /** Adds the item to the queue for writing. */
+    def write(item: T): Unit = this.add(item)
+  }
+
+  /** Asynchronously writes to multiple writers.  Groups the writers into `ceil(N, parallelism)` where `N` is the number
+    * of writers.  Creates a thread for each group to asynchronously write all elements that are written to writers
+    * in that group.
+    *
+    * @param writers the writers to which to write
+    * @param bufferSize the number of elements to buffer for each group of writers, before we block writing to writers
+    *                   in that group
+    * @param parallelism the maximum parallelism to use
+    * @param toIndex a method that returns the index into `writers` for each element to be written
+    * @tparam T the type of object to write
+    */
+  class AsyncGroupingWriter[T](writers: Traversable[Writer[T]],
```

Comment:
I don't really like this model.  It has a bunch of shortcomings, not least of which are:

1. The parallelism is silod.  E.g. if you have 96 writers, and one partition or group is getting 90% of the data, you're not going to see much benefit.
2. It requires you to be able to determine the writer from the thing being written which seems very task-specific

I think for something like this to live in commons we probably want it to be more general, and not have bottlenecks based on how data is distributed among the writers.